tests/functional/test_s3.py

```diff
@@ -790,3 +792,113 @@ def _verify_expected_endpoint_url(region, bucket, key, s3_config,
+    # We special case presigned URLs due to backwards
+    # compatibility.
+    t = S3AddressingCases(_verify_presigned_url_addressing)
+
+    # us-east-1, or the "global" endpoint. A signature version of
+    # None means the user doesn't have signature version configured.
+    yield t.case(region='us-east-1', bucket='bucket', key='key',
+                 signature_version=None,
+                 expected_url='https://bucket.s3.amazonaws.com/key')
+    yield t.case(region='us-east-1', bucket='bucket', key='key',
+                 signature_version='s3',
+                 expected_url='https://bucket.s3.amazonaws.com/key')
+    yield t.case(region='us-east-1', bucket='bucket', key='key',
+                 signature_version='s3v4',
+                 expected_url='https://bucket.s3.amazonaws.com/key')
+    yield t.case(region='us-east-1', bucket='bucket', key='key',
+                 signature_version='s3v4',
+                 s3_config={'addressing_style': 'path'},
+                 expected_url='https://s3.amazonaws.com/bucket/key')
+
+    # A region that supports both 's3' and 's3v4'.
+    yield t.case(region='us-west-2', bucket='bucket', key='key',
+                 signature_version=None,
+                 expected_url='https://bucket.s3.amazonaws.com/key')
+    yield t.case(region='us-west-2', bucket='bucket', key='key',
+                 signature_version='s3',
+                 expected_url='https://bucket.s3.amazonaws.com/key')
+    yield t.case(region='us-west-2', bucket='bucket', key='key',
+                 signature_version='s3v4',
+                 expected_url='https://bucket.s3.amazonaws.com/key')
+    yield t.case(region='us-west-2', bucket='bucket', key='key',
+                 signature_version='s3v4',
+                 s3_config={'addressing_style': 'path'},
+                 expected_url='https://s3.us-west-2.amazonaws.com/bucket/key')
+
+    # An 's3v4' only region.
+    yield t.case(region='us-east-2', bucket='bucket', key='key',
+                 signature_version=None,
+                 expected_url='https://bucket.s3.amazonaws.com/key')
+    yield t.case(region='us-east-2', bucket='bucket', key='key',
+                 signature_version='s3',
+                 expected_url='https://bucket.s3.amazonaws.com/key')
+    yield t.case(region='us-east-2', bucket='bucket', key='key',
+                 signature_version='s3v4',
+                 expected_url='https://bucket.s3.amazonaws.com/key')
+    yield t.case(region='us-east-2', bucket='bucket', key='key',
+                 signature_version='s3v4',
+                 s3_config={'addressing_style': 'path'},
+                 expected_url='https://s3.us-east-2.amazonaws.com/bucket/key')
+
+    # Dualstack endpoints
+    yield t.case(region='us-west-2', bucket='bucket', key='key',
+                 signature_version=None,
+                 s3_config={'use_dualstack_endpoint': True},
+                 expected_url='https://bucket.s3.amazonaws.com/key')
```

Comment:
Hmm the expected url does not look like a dualstack endpoint. Should that be the case for this?

Reply:
Yeah seems fine with me.